### PR TITLE
Warn if setq-local is used without emacs 24.3

### PR DESCRIPTION
**Pull request by omajid**
This adds a warning when `setq-local` is used without requiring emacs 24.3 as suggested [here]( https://github.com/milkypostman/melpa/pull/2631#issuecomment-87274087).

It has some flaws:

- Only activates if a Package-Requires header is present
- It looks for the literal text `setq-local`, even if it is commented out.

### CODE REVIEW

**Comment by purcell** [comment]
Thanks for taking the time to submit this.

Yes, I'd like to do this in a slightly more general way: there are a bunch of functions which are new (e.g. `defvar-local`), and also newly-added libraries. I'd really like to have a lookup table of new functions, variables and libraries, mapping them to the corresponding Emacs version which introduced them -- then we'd be able to handle this in a more general way.

I'll look at this patch in more detail tomorrow and may well merge it pending later generalisations -- thanks again!

**Comment by Fanael** [comment]
NAK from me: it warns on
```emacs-lisp
;; We'd use `setq-local' here, but it's Emacs 24.3+ only
(set (make-local-variable 'foo) 'bar)
```

**Comment by omajid** [comment]
> we should really walk the code

Any suggestions on how to implement it?

**Comment by omajid** [comment]
This new version deals correctly with both comments and strings.

    ;; avoid using setq-local
    ;; (setq-local 'foo 10)
    (setq-local 'foo-bar 10)
    (let ((foo-bar "setq-local")))
    (setq-local 'foo-baz 20)

Only lines 3 and 5 are highlighted as errors.

**Comment by purcell** [comment]
Thanks - this looks good. It would currently match `my-setq-local` or `setq-localized`, so perhaps better to match `(setq-local` followed by a word boundary? Once done, I'm keen to merge this and then generalise it a little to easily handle some common similar cases, just as `defvar-local`.

**Comment by omajid** [comment]
The patch uses `(string-equal (symbol-at-point) ...)`. In my tests, it doesn't match `setq-localized` or `my-setq-local`. Did I overlook something?

File: flycheck-package.el [hunk 1]
```diff
@@ -208,6 +209,25 @@ the form (PACKAGE-NAME PACKAGE-VERSION LINE-NO LINE-BEGINNING-OFFSET)."
          lexbind-line lexbind-col 'warning
          "You should depend on (emacs \"24\") if you need lexical-binding.")))))
 
+(defun flycheck-package--check-setq-local-requires-emacs-24.3 (valid-deps)
+  "Warn about use of `setq-local' when Emacs 24.3 is not among VALID-DEPS."
+  (goto-char (point-min))
+  (while (re-search-forward "setq-local" (point-max) t)
+    (when (and (string-equal (symbol-at-point) "setq-local")
```
Comment (Fanael) [inline flycheck-package.el]: You could search for `(rx symbol-start "setq-local" symbol-end)` instead, so there'd be no need for `symbol-at-point`.

File: flycheck-package.el [hunk 2]
```diff
@@ -208,6 +209,25 @@ the form (PACKAGE-NAME PACKAGE-VERSION LINE-NO LINE-BEGINNING-OFFSET)."
...
+               (null (nth 3 (syntax-ppss)))     ;; not a string
+               (null (nth 4 (syntax-ppss))))    ;; not a comment
+      (let* ((emacs-dep (assq 'emacs valid-deps))
+             (emacs-version (if emacs-dep
+                                (cadr emacs-dep)
+                              nil)))
```
Comment (Fanael) [inline flycheck-package.el]: `cadr` returns `nil` when called on `nil`, so `(let ((emacs-version (cadr (assq 'emacs valid-deps)))) …)` would be enough.

File: flycheck-package.el [hunk 3]
```diff
@@ -208,6 +209,25 @@ the form (PACKAGE-NAME PACKAGE-VERSION LINE-NO LINE-BEGINNING-OFFSET)."
...
+        (when (or (eq emacs-version nil)
```
Comment (Fanael) [inline flycheck-package.el]: `null`

**Comment by Fanael** [comment]
> so perhaps better to match `(setq-local`

This approach is better, as it doesn't warn on `(fboundp 'setq-local)`.

**Comment by omajid** [comment]
True. On the flip side, my current approach warns on `(apply #'setq-local ...)`. I will change it as you suggest, though.

**Comment by Fanael** [comment]
> On the flip side, my current approach warns on `(apply #'setq-local ...)`

That's true, but `setq-local` is a macro, so that doesn't even work.

**Comment by omajid** [comment]
I am an idiot :)

**Comment by Fanael** [comment]
Merged and extended to detect `defvar-local`, too.

**Comment by omajid** [comment]
Thanks!

**Pull request merged**